localstack/services/kinesisfirehose/resource_providers/aws_kinesisfirehose_deliverystream.py

```diff
@@ -0,0 +1,492 @@
+          - firehose:DescribeDeliveryStream
+          - iam:GetRole
+          - iam:PassRole
+          - kms:CreateGrant
+          - kms:DescribeKey
+
+        """
+        model = request.desired_state
+        firehose = request.aws_client_factory.firehose
+        parameters = [
+            "DeliveryStreamName",
+            "DeliveryStreamType",
+            "S3DestinationConfiguration",
+            "ElasticsearchDestinationConfiguration",
+            "AmazonopensearchserviceDestinationConfiguration",
+            "DeliveryStreamEncryptionConfigurationInput",
+            "ExtendedS3DestinationConfiguration",
+            "HttpEndpointDestinationConfiguration",
+            "KinesisStreamSourceConfiguration",
+            "RedshiftDestinationConfiguration",
+            "SplunkDestinationConfiguration",
+            "Tags",
+        ]
+        attrs = util.select_attributes(model, params=parameters)
+
+        if not request.custom_context.get(REPEATED_INVOCATION):
+            response = firehose.create_delivery_stream(**attrs)
+            # TODO: defaults
+            # TODO: idempotency
+            model["Arn"] = response["DeliveryStreamARN"]
+            request.custom_context[REPEATED_INVOCATION] = True
+            return ProgressEvent(
+                status=OperationStatus.IN_PROGRESS,
+                resource_model=model,
+                custom_context=request.custom_context,
+            )
+
+        stream = firehose.describe_delivery_stream(DeliveryStreamName=model["DeliveryStreamName"])
+        if stream["DeliveryStreamDescription"]["DeliveryStreamStatus"] != "ACTIVE":
+            return ProgressEvent(
+                status=OperationStatus.IN_PROGRESS,
+                resource_model=model,
+                custom_context=request.custom_context,
+            )
+
+        return ProgressEvent(
+            status=OperationStatus.SUCCESS,
+            resource_model=model,
+            custom_context=request.custom_context,
+        )
+
+    def read(
+        self,
+        request: ResourceRequest[KinesisFirehoseDeliveryStreamProperties],
+    ) -> ProgressEvent[KinesisFirehoseDeliveryStreamProperties]:
+        """
+        Fetch resource information
+
+        IAM permissions required:
+          - firehose:DescribeDeliveryStream
+          - firehose:ListTagsForDeliveryStream
+        """
+        raise NotImplementedError
+
+    def delete(
+        self,
+        request: ResourceRequest[KinesisFirehoseDeliveryStreamProperties],
+    ) -> ProgressEvent[KinesisFirehoseDeliveryStreamProperties]:
+        """
+        Delete a resource
+
+        IAM permissions required:
+          - firehose:DeleteDeliveryStream
+          - firehose:DescribeDeliveryStream
+          - kms:RevokeGrant
+          - kms:DescribeKey
+        """
+        model = request.desired_state
+        firehose = request.aws_client_factory.firehose
+        firehose.delete_delivery_stream(DeliveryStreamName=model["DeliveryStreamName"])
```

Comment:
Question: does the delivery stream on AWS take a while to delete? Should we use the `REPEATED_INVOCATION` pattern here as well?

Reply:
It does take some time to delete and while in DELETING state it might still accept info. I'd have to measure it to see how long it take in AWS.

From boto3 documentation
```While the delivery stream is in the DELETING state, the service might continue to accept records, but it doesn’t make any guarantees with respect to delivering the data. Therefore, as a best practice, first stop any applications that are sending records before you delete a delivery stream.```
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/firehose/client/delete_delivery_stream.html